rest/blip_api_crud_test.go

```diff
@@ -3151,3 +3151,92 @@ func TestOnDemandImportBlipFailure(t *testing.T) {
 		}
 	})
 }
+
+// TestBlipDatabaseClose verifies that the client connection is closed when the database is closed.
+// Starts a continuous pull replication then updates the db to trigger a close.
+func TestBlipDatabaseClose(t *testing.T) {
+
```

Comment:
I think there's a pretty good argument for setting this up with `btcRunner := NewBlipTesterClientRunner(t)` which will parameterize using 3.x and 4.x messages. Using `btcRunner.NewBlipTesterClientOptsWithRT(rt, opts)` would allow you to not have to worry about implementing handlers, and you can do `btcRunner.StartPullSince`.

I think this would be equivalent to the test below but with more common boilerplate.

I think there's a different argument to be made that /db/_blipsync should be called, since requests contexts are at place here, and testing for the actual error status to see that we don't regress. I don't think we yet have good boilerplate for this because while ISGR lets you use a real URL https://github.com/couchbase/sync_gateway/blob/08686266a764edab6df8dd0361913af00d60be1d/rest/replicatortest/replicator_test.go#L7793 we don't have an easy way to send blip requests.

Reply:
I think it's fine to not set up real _blipsync here, we can be sure it exits with some error code. Here's some more up to date boilerplate for this test:

```
unc TestBlipDatabaseClose(t *testing.T) {

	base.SetUpTestLogging(t, base.LevelInfo, base.KeyHTTP, base.KeySync, base.KeySyncMsg, base.KeyChanges, base.KeyCache)
	btcRunner := NewBlipTesterClientRunner(t)
	btcRunner.Run(func(t *testing.T, SupportedBLIPProtocols []string) {
		rt := NewRestTesterPersistentConfig(t)
		defer rt.Close()
		const username = "alice"
		rt.CreateUser(username, []string{"*"})
		btc := btcRunner.NewBlipTesterClientOptsWithRT(rt, &BlipTesterClientOpts{Username: username})
		blipContextClosed := false
		btcRunner.clients[btc.id].pullReplication.bt.blipContext.OnExitCallback = func() {
			blipContextClosed = true
		}

		// put a doc, and make sure blip connection is established
		markerDoc := "markerDoc"
		markerDocVersion := rt.CreateTestDoc(markerDoc)
		rt.WaitForPendingChanges()
		btcRunner.StartPull(btc.id)

		btcRunner.WaitForVersion(btc.id, markerDoc, markerDocVersion)

		RequireStatus(t, rt.SendAdminRequest(http.MethodDelete, "/{{.db}}/", ""), http.StatusOK)

		require.EventuallyWithT(t, func(c *assert.CollectT) {
			assert.True(c, blipContextClosed)
		}, time.Second*10, time.Millisecond*100)
	})
}
```

Reply:
The test-race failure is from `blipContextClosed` not being an atomic.